File: Sources/NotchAgent/Features/Providers/Claude/ClaudeProvider.swift
import Foundation

/// Claude Code usage from two complementary sources:
/// 1. Local transcripts (`~/.claude/projects/**/*.jsonl`) — tokens, cost
///    estimates, 5h session blocks, hourly activity. Always available.
/// 2. Optional API probe (`ClaudeQuotaProbe`) — authoritative 5h/7d quota
///    percentages, reset times and limit status from response headers.
/// When the probe is disabled or no OAuth token exists, percentages fall back
/// to user-configured budgets (or stay hidden — never fabricated).
struct ClaudeProvider: UsageProvider {
    let id = ProviderID.claudeCode
    let capabilities: ProviderCapabilities = [
        .sessionTokens, .sessionPercent, .weeklyTokens, .weeklyPercent, .costEstimate, .resetSchedule,
    ]

    private let roots: [URL]
    private let cache = ClaudeScanCache()
    private let probe: ClaudeQuotaProbe?
    private static let lookback: TimeInterval = 8 * 24 * 3600

    /// Every place Claude Code writes transcripts on this Mac: the CLI and
    /// the Desktop app's agent mode (same JSONL format, different root).
    static let defaultRoots: [URL] = [
        AppPaths.home.appendingPathComponent(".claude/projects"),
        AppPaths.home.appendingPathComponent("Library/Application Support/Claude/local-agent-mode-sessions"),
    ]

    init(roots: [URL] = ClaudeProvider.defaultRoots, probe: ClaudeQuotaProbe? = ClaudeQuotaProbe()) {
        self.roots = roots
        self.probe = probe
    }

    init(root: URL, probe: ClaudeQuotaProbe?) {
        self.init(roots: [root], probe: probe)
    }

    func detectInstallation() -> ProviderInstallation {
        for root in roots where FileManager.default.fileExists(atPath: root.path) {
            return .installed(dataPath: root.path)
        }
        return .notInstalled
    }

    func fetchSnapshot(settings: AppSettings) async throws -> UsageSnapshot {
        let now = Date()
        guard case .installed = detectInstallation() else {
            return UsageSnapshot(provider: id, health: .notInstalled)
        }

        // Freshness rules (review finding): a cached quota is only trusted
        // while it is recent AND its windows have not already reset — an
        // expired token or dead network must never freeze yesterday's 60%
        // as today's truth. Stale quota degrades to the local heuristics.
        var quota: ClaudeQuota?
        if settings.claudeQuotaProbeEnabled, let probe {
            quota = await probe.currentQuota()
            if let fetched = quota?.fetchedAt, now.timeIntervalSince(fetched) > 15 * 60 {
                quota = nil
            }
        }
        let freshSessionReset = quota?.sessionResetsAt.flatMap { $0 > now ? $0 : nil }
        let freshWeeklyReset = quota?.weeklyResetsAt.flatMap { $0 > now ? $0 : nil }

        let files = roots.flatMap {
            recentFiles(under: $0, ext: "jsonl", modifiedAfter: now.addingTimeInterval(-Self.lookback))
        }
        var merged: [Date: ClaudeFileStat.HourStat] = [:]
        var mergedModels: [String: ClaudeFileStat.ModelStat] = [:]
        var lastActivity: Date?
        var lastModel: String?
        var failedFiles = 0

        for url in files {
            let stat: ClaudeFileStat?
            do {
                stat = try await cache.stat(for: url)
            } catch {
                failedFiles += 1
                Log.providers.error("claude: failed to parse \(url.lastPathComponent, privacy: .public): \(error.localizedDescription, privacy: .public)")
                continue
            }
            guard let stat else { continue }

            for (hour, bucket) in stat.hours {
                var existing = merged[hour] ?? .init()
                existing.tokens += bucket.tokens
                existing.costUSD += bucket.costUSD
                existing.messages += bucket.messages
                merged[hour] = existing
            }
            for (model, modelStat) in stat.byModel {
                var existing = mergedModels[model] ?? .init()
                existing.tokens += modelStat.tokens
                existing.costUSD += modelStat.costUSD
                mergedModels[model] = existing
            }
            if let activity = stat.lastActivity, lastActivity.map({ activity > $0 }) ?? true {
                lastActivity = activity
                lastModel = stat.lastModel
            }
        }

        guard !merged.isEmpty || quota != nil else {
            let health: ProviderHealth = failedFiles > 0 ? .parseError : .noData
            return UsageSnapshot(provider: id, health: health, note: files.isEmpty ? "No activity in the last 8 days" : nil)
        }

        // Session window: the API reset time is authoritative (start = reset − 5h,
        // matching what the % refers to); the local block heuristic is fallback.
        var session: SessionUsage?
        let sessionWindow: (start: Date, end: Date)? =
            freshSessionReset.map { ($0.addingTimeInterval(-SessionBlocks.blockLength), $0) }
            ?? SessionBlocks.currentBlock(activityHours: Array(merged.keys), now: now)
        if let window = sessionWindow {
            let (tokens, cost) = Self.sumBuckets(merged, from: window.start, to: window.end)
            session = SessionUsage(
                tokens: tokens,
                cost: CostEstimate(amountUSD: cost),
                startedAt: window.start,
                resetsAt: window.end,
                usedPercent: (freshSessionReset != nil ? quota?.sessionPercent : nil)
                    ?? settings.claudeSessionTokenBudget.map { budget in
                        min(100, Double(tokens.total) / Double(max(budget, 1)) * 100)
                    }
            )
        }

        // Weekly window aligned to the API reset when known (tokens must refer
        // to the same window as the percentage beside them).
        let weekCutoff = freshWeeklyReset.map { $0.addingTimeInterval(-7 * 24 * 3600) }
            ?? now.addingTimeInterval(-7 * 24 * 3600)
        var weekTokens = TokenUsage.zero
        var weekCost = 0.0
        var byDay: [Date: (tokens: Int, cost: Double)] = [:]
        var hourly: [HourlyTotal] = []
        for (hour, bucket) in merged where hour >= weekCutoff {
            weekTokens += bucket.tokens
            weekCost += bucket.costUSD
            let day = hour.flooredToDay
            let current = byDay[day] ?? (0, 0)
            byDay[day] = (current.tokens + bucket.tokens.total, current.cost + bucket.costUSD)
            hourly.append(HourlyTotal(hour: hour, tokens: bucket.tokens.total))
        }
        let weekly = WeeklyUsage(
            tokens: weekTokens,
            cost: CostEstimate(amountUSD: weekCost),
            usedPercent: (freshWeeklyReset != nil ? quota?.weeklyPercent : nil)
                ?? settings.claudeWeeklyTokenBudget.map { budget in
                    min(100, Double(weekTokens.total) / Double(max(budget, 1)) * 100)
                },
            resetsAt: freshWeeklyReset,
            dailyTotals: byDay
                .map { DailyTotal(day: $0.key, tokens: $0.value.tokens, costUSD: $0.value.cost) }
                .sorted { $0.day < $1.day },
            hourlyTotals: hourly.sorted { $0.hour < $1.hour }
        )

        let breakdown = mergedModels
            .map { ModelUsage(model: $0.key, tokens: $0.value.tokens.total, costUSD: $0.value.costUSD) }
            .sorted { $0.costUSD > $1.costUSD }

        var modelHealth: [ModelHealth]?
        if settings.claudeQuotaProbeEnabled, let probe {
            let recorded = await probe.modelHealthSnapshot()
            modelHealth = recorded.isEmpty ? nil : recorded
        }

        // Evict cache entries for files that left the lookback window —
        // without this, seenKeys of every transcript ever seen stay resident.
        await cache.prune(keeping: Set(files.map(\.path)))

        return UsageSnapshot(
            provider: id,
            capturedAt: now,
            health: failedFiles > 0 ? .degraded : .ok,
            session: session,
            weekly: weekly,
            activeModel: lastModel,
            lastActivityAt: lastActivity,
            note: limitingNote(quota),
            quotaStatus: freshSessionReset != nil || freshWeeklyReset != nil ? quota?.status : nil,
            modelBreakdown: breakdown.isEmpty ? nil : breakdown,
            modelHealth: modelHealth
        )
    }

    /// Hour buckets are the finest grain we keep, so window boundaries carry
    /// up to ±1 bucket of imprecision — documented in the README.
    static func sumBuckets(
        _ merged: [Date: ClaudeFileStat.HourStat],
        from start: Date,
        to end: Date
    ) -> (tokens: TokenUsage, costUSD: Double) {
        var tokens = TokenUsage.zero
        var cost = 0.0
        let flooredStart = start.flooredToHour
        for (hour, bucket) in merged where hour >= flooredStart && hour < end {
            tokens += bucket.tokens
            cost += bucket.costUSD
        }
        return (tokens, cost)
    }

    private func limitingNote(_ quota: ClaudeQuota?) -> String? {
        switch quota?.limitingWindow {
        case "five_hour": "Limiting window: 5h session"
        case "seven_day": "Limiting window: 7-day"
        default: nil
        }
    }
}

File: windows/NotchAgent.Windows/Providers/Claude/ClaudeProvider.cs
using NotchAgent.Windows.Models;
using NotchAgent.Windows.Providers.Shared;
using NotchAgent.Windows.Services;

namespace NotchAgent.Windows.Providers.Claude;

/// Claude Code usage from two complementary sources:
/// 1. Local transcripts — tokens, cost estimates, 5h session blocks, hourly
///    activity. Always available. Scans both the CLI's own transcripts and
///    (when present) the Desktop app's agent-mode sessions.
/// 2. Optional API probe — authoritative 5h/7d quota percentages, reset
///    times and limit status from response headers.
public sealed class ClaudeProvider : IUsageProvider
{
    public ProviderId Id => ProviderId.ClaudeCode;

    private readonly List<string> _roots;
    private readonly ClaudeScanCache _cache = new();
    private readonly ClaudeQuotaProbe? _probe;
    private static readonly TimeSpan Lookback = TimeSpan.FromDays(8);
    private static readonly TimeSpan BlockLength = TimeSpan.FromHours(5);

    public ClaudeProvider(IEnumerable<string>? roots = null, ClaudeQuotaProbe? probe = null)
    {
        _roots = roots?.ToList() ?? DefaultRoots();
        _probe = probe ?? new ClaudeQuotaProbe();
    }

    private static List<string> DefaultRoots() => new()
    {
        Path.Combine(AppPaths.Home, ".claude", "projects"),
    };

    public ProviderInstallation DetectInstallation()
    {
        foreach (var root in _roots)
        {
            if (Directory.Exists(root)) return ProviderInstallation.Installed(root);
        }
        return ProviderInstallation.NotInstalled;
    }

    public async Task<UsageSnapshot> FetchSnapshotAsync(AppSettings settings, CancellationToken ct)
    {
        var now = DateTimeOffset.UtcNow;
        if (DetectInstallation().Kind != ProviderInstallationKind.Installed)
        {
            return new UsageSnapshot { Provider = Id, Health = ProviderHealth.NotInstalled };
        }

        // Freshness rules: a cached quota is only trusted while it is recent
        // AND its windows have not already reset — an expired token or dead
        // network must never freeze yesterday's 60% as today's truth.
        ClaudeQuota? quota = null;
        if (settings.ClaudeQuotaProbeEnabled && _probe is not null)
        {
            quota = await _probe.CurrentQuotaAsync(ct);
            if (quota is { FetchedAt: var fetched } && now - fetched > TimeSpan.FromMinutes(15))
            {
                quota = null;
            }
        }
        var freshSessionReset = quota?.SessionResetsAt is { } sr && sr > now ? sr : (DateTimeOffset?)null;
        var freshWeeklyReset = quota?.WeeklyResetsAt is { } wr && wr > now ? wr : (DateTimeOffset?)null;

        var files = _roots.SelectMany(r => FileScan.RecentFiles(r, "jsonl", now - Lookback)).ToList();
        var merged = new Dictionary<DateTimeOffset, ClaudeFileStat.HourStat>();
        var mergedModels = new Dictionary<string, ClaudeFileStat.ModelStat>();
        DateTimeOffset? lastActivity = null;
        string? lastModel = null;
        int failedFiles = 0;

        foreach (var url in files)
        {
            ClaudeFileStat? stat;
            try
            {
                stat = _cache.Stat(url);
            }
            catch (IOException ex)
            {
                failedFiles++;
                Log.Providers.LogError("claude: failed to parse {0}: {1}", Path.GetFileName(url), ex.Message);
                continue;
            }
            if (stat is null) continue;

            foreach (var (hour, bucket) in stat.Hours)
            {
                if (!merged.TryGetValue(hour, out var existing))
                {
                    existing = new ClaudeFileStat.HourStat();
                    merged[hour] = existing;
                }
                existing.Tokens += bucket.Tokens;
                existing.CostUsd += bucket.CostUsd;
                existing.Messages += bucket.Messages;
            }
            foreach (var (model, modelStat) in stat.ByModel)
            {
                if (!mergedModels.TryGetValue(model, out var existing))
                {
                    existing = new ClaudeFileStat.ModelStat();
                    mergedModels[model] = existing;
                }
                existing.Tokens += modelStat.Tokens;
                existing.CostUsd += modelStat.CostUsd;
            }
            if (stat.LastActivity is { } activity && (lastActivity is null || activity > lastActivity))
            {
                lastActivity = activity;
                lastModel = stat.LastModel;
            }
        }

        var scannedPaths = files.ToHashSet();
        _cache.Prune(scannedPaths);

        if (merged.Count == 0 && quota is null)
        {
            var health = failedFiles > 0 ? ProviderHealth.ParseError : ProviderHealth.NoData;
            return new UsageSnapshot
            {
                Provider = Id,
                Health = health,
                Note = files.Count == 0 ? "No activity in the last 8 days" : null,
            };
        }

        // Session window: the API reset time is authoritative (start = reset
        // − 5h, matching what the % refers to); the local block heuristic is
        // fallback.
        SessionUsage? session = null;
        (DateTimeOffset Start, DateTimeOffset End)? sessionWindow =
            freshSessionReset is { } fsr ? (fsr - BlockLength, fsr) : CurrentBlock(merged.Keys, now);
        if (sessionWindow is { } window)
        {
            var (tokens, cost) = SumBuckets(merged, window.Start, window.End);
            double? usedPercent = freshSessionReset is not null ? quota?.SessionPercent : null;
            usedPercent ??= settings.ClaudeSessionTokenBudget is { } budget
                ? Math.Min(100, (double)tokens.Total / Math.Max(budget, 1) * 100)
                : null;
            session = new SessionUsage
            {
                Tokens = tokens,
                Cost = new CostEstimate { AmountUsd = cost },
                StartedAt = window.Start,
                ResetsAt = window.End,
                UsedPercent = usedPercent,
            };
        }

        // Weekly window aligned to the API reset when known.
        var weekCutoff = freshWeeklyReset is { } fwr ? fwr - TimeSpan.FromDays(7) : now - TimeSpan.FromDays(7);
        var weekTokens = new TokenUsage();
        double weekCost = 0;
        var byDay = new Dictionary<DateTimeOffset, (long Tokens, double Cost)>();
        var hourly = new List<HourlyTotal>();
        foreach (var (hour, bucket) in merged)
        {
            if (hour < weekCutoff) continue;
            weekTokens += bucket.Tokens;
            weekCost += bucket.CostUsd;
            var day = new DateTimeOffset(hour.Date, TimeSpan.Zero);
            var current = byDay.GetValueOrDefault(day, (0, 0));
            byDay[day] = (current.Tokens + bucket.Tokens.Total, current.Cost + bucket.CostUsd);
            hourly.Add(new HourlyTotal { Hour = hour, Tokens = bucket.Tokens.Total });
        }

        double? weeklyPercent = freshWeeklyReset is not null ? quota?.WeeklyPercent : null;
        weeklyPercent ??= settings.ClaudeWeeklyTokenBudget is { } weekBudget
            ? Math.Min(100, (double)weekTokens.Total / Math.Max(weekBudget, 1) * 100)
            : null;

        var weekly = new WeeklyUsage
        {
            Tokens = weekTokens,
            Cost = new CostEstimate { AmountUsd = weekCost },
            UsedPercent = weeklyPercent,
            ResetsAt = freshWeeklyReset,
            DailyTotals = byDay.Select(kv => new DailyTotal { Day = kv.Key, Tokens = kv.Value.Tokens, CostUsd = kv.Value.Cost })
                .OrderBy(d => d.Day).ToList(),
            HourlyTotals = hourly.OrderBy(h => h.Hour).ToList(),
        };

        var breakdown = mergedModels
            .Select(kv => new ModelUsage { Model = kv.Key, Tokens = kv.Value.Tokens.Total, CostUsd = kv.Value.CostUsd })
            .OrderByDescending(m => m.CostUsd)
            .ToList();

        var quotaStatus = (freshSessionReset is not null || freshWeeklyReset is not null)
            ? quota?.Status switch
            {
                ClaudeQuotaStatus.Ok => Models.QuotaStatus.Ok,
                ClaudeQuotaStatus.Warning => Models.QuotaStatus.Warning,
                ClaudeQuotaStatus.Blocked => Models.QuotaStatus.Blocked,
                _ => (Models.QuotaStatus?)null,
            }
            : null;

        return new UsageSnapshot
        {
            Provider = Id,
            CapturedAt = now,
            Health = failedFiles > 0 ? ProviderHealth.Degraded : ProviderHealth.Ok,
            Session = session,
            Weekly = weekly,
            ActiveModel = lastModel,
            LastActivityAt = lastActivity,
            Note = LimitingNote(quota),
            QuotaStatus = quotaStatus,
            ModelBreakdown = breakdown.Count > 0 ? breakdown : null,
        };
    }

    /// Claude Plans meter usage in rolling ~5h session blocks: a block starts
    /// at the floored hour of the first activity after the previous block
    /// ended, and lasts 5 hours.
    private static (DateTimeOffset Start, DateTimeOffset End)? CurrentBlock(IEnumerable<DateTimeOffset> activityHours, DateTimeOffset now)
    {
        DateTimeOffset? blockStart = null;
        var blockEnd = DateTimeOffset.MinValue;
        foreach (var hour in activityHours.OrderBy(h => h))
        {
            if (blockStart is null || hour >= blockEnd)
            {
                blockStart = hour;
                blockEnd = hour + BlockLength;
            }
        }
        if (blockStart is not { } start || now < start || now >= blockEnd) return null;
        return (start, blockEnd);
    }

    /// Hour buckets are the finest grain we keep, so window boundaries carry
    /// up to ±1 bucket of imprecision — documented in the README.
    private static (TokenUsage Tokens, double CostUsd) SumBuckets(
        Dictionary<DateTimeOffset, ClaudeFileStat.HourStat> merged, DateTimeOffset start, DateTimeOffset end)
    {
        var tokens = new TokenUsage();
        double cost = 0;
        var flooredStart = ClaudeTranscriptParser.FlooredToHour(start);
        foreach (var (hour, bucket) in merged)
        {
            if (hour >= flooredStart && hour < end)
            {
                tokens += bucket.Tokens;
                cost += bucket.CostUsd;
            }
        }
        return (tokens, cost);
    }

    private static string? LimitingNote(ClaudeQuota? quota) => quota?.LimitingWindow switch
    {
        "five_hour" => "Limiting window: 5h session",
        "seven_day" => "Limiting window: 7-day",
        _ => null,
    };
}
